server/src/main/java/io/unitycatalog/server/service/iceberg/MetadataService.java

```diff
@@ -0,0 +1,35 @@
+package io.unitycatalog.server.service.iceberg;
+
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableMetadataParser;
+import org.apache.iceberg.io.FileIO;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class MetadataService {
+
+  private final FileIOFactory fileIOFactory;
+
+  public MetadataService(FileIOFactory fileIOFactory) {
+    this.fileIOFactory = fileIOFactory;
+  }
+
+  public TableMetadata readTableMetadata(String metadataLocation) throws IOException {
+    URI metadataLocationUri = URI.create(metadataLocation);
+    // TODO: cache fileIO
+    FileIO fileIO = fileIOFactory.getFileIO(metadataLocationUri);
+
+    TableMetadata tableMetadata;
+    if (fileIO == null) {
```

Comment:
This fallback logic looks quite ad-hoc. Can we define a subclass of FileIO that handles the local file system read, and return it in the FileIOFactory?

Reply:
Seemed a little overkill to write a FileIO impl just to read a local metadata file. We can however fallback to HadoopFileIO (which handles local filesystem) -- this just introduces a dependency on hadoop-common

Reply:
okay, well I added a very simple fileIO impl but I'm not sure this should be the long term solution and added a TODO about it